core/src/main/java/org/apache/iceberg/io/ClusteredDeleteWriter.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.List;
+import org.apache.iceberg.DeleteFile;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.util.CharSequenceSet;
+
+public abstract class ClusteredDeleteWriter<T> extends ClusteredWriter<T, DeleteWriteResult> {
+
+  private final List<DeleteFile> deleteFiles = Lists.newArrayList();
+  private final CharSequenceSet referencedDataFiles = CharSequenceSet.empty();
+
+  @Override
+  protected void add(DeleteWriteResult result) {
+    deleteFiles.addAll(result.deleteFiles());
+    referencedDataFiles.addAll(result.referencedDataFiles());
```

Comment:
I noticed `CharSequenceSet#addAll` calls in flame graphs (nothing too bad). If we want to squeeze a little bit of performance, we can replace `addAll` with `union` to avoid coping things. Alternatively, we can detect whether whether the set we add is also `CharSequenceSet`. If so, no need to unwrap and wrap records again.


I am not sure it is going to be worth it, though.